From 0bd9858b7b75c822adc8d439ec6547d4c2089f86 Mon Sep 17 00:00:00 2001
From: Ossama Othman <ossama.othman@intel.com>
Date: Fri, 22 Apr 2016 13:43:14 -0700
Subject: [PATCH 1/2] mptcp: prevent subflows on master interface from being
 recreated

The subflows for the master interface in the fullmesh_mf path manager
should only be created once.  Reintroduce logic found in the original
fullmesh path manager that prevented subflows for the master socket
from being recreated.  This addresses a crash that would occur after
the master interface was brought down then up.

Change-Id: Iceeac92ced3d9531e12672b6fa2a58ff79d15780
Signed-off-by: Ossama Othman <ossama.othman@intel.com>
---
 net/mptcp/mptcp_fullmesh_mf.c | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/net/mptcp/mptcp_fullmesh_mf.c b/net/mptcp/mptcp_fullmesh_mf.c
--- a/net/mptcp/mptcp_fullmesh_mf.c
+++ b/net/mptcp/mptcp_fullmesh_mf.c
@@ -89,6 +89,9 @@ struct fullmesh_priv {
 
 	u8 rem4_bits;
 	u8 rem6_bits;
+
+	/* Are we established the additional subflows for primary pair? */
+	u8 first_pair:1;
 };
 
 struct mptcp_fm_ns {
@@ -486,7 +489,7 @@ next_subflow:
 	lock_sock_nested(meta_sk, SINGLE_DEPTH_NESTING);
 
 	/* Create the additional subflows for the master interface */
-	if (master_tp->mptcp->master) {
+	if (master_tp && master_tp->mptcp->master && fmp->first_pair == 0) {
 		struct mptcp_loc4 loc;
 		struct mptcp_rem4 rem;
 
@@ -501,6 +504,8 @@ next_subflow:
 
 		mptcp_debug("%s creating additional IPv4 subflows for the master interface\n", __func__);
 		mptcp_v4_subflows(meta_sk, &loc, &rem);
+
+		fmp->first_pair = 1;
 	}
 	iter++;
 
@@ -542,7 +547,7 @@ next_subflow:
 	}
 
 #if IS_ENABLED(CONFIG_IPV6)
-	if (master_tp->mptcp->master) {
+	if (master_tp && master_tp->mptcp->master && fmp->first_pair == 0) {
 			struct mptcp_loc6 loc;
 			struct mptcp_rem6 rem;
 
@@ -557,6 +562,8 @@ next_subflow:
 
 			mptcp_debug("%s creating additional IPv6 subflows for the master interface\n", __func__);
 			mptcp_v6_subflows(meta_sk, &loc, &rem);
+
+			fmp->first_pair = 1;
 	}
 	mptcp_for_each_bit_set(fmp->rem6_bits, i) {
 		struct fullmesh_rem6 *rem;
